### ISSUE TITLE
Guile plugin crashing when exiting WeeChat

### ISSUE THREAD

**Issue by drew-kun** [weechat/weechat#527]
Hi!
I've installed weechat v1.3 on OS X 10.10.5 from homebrew with all this options:

```
weechat --with-aspell --with-curl --with-guile --with-lua --with-perl --with-python --with-ruby
```

Every single time when I exit weechat `/q` or `/exit` or `quit` it returns me this message:

```
*** Very bad! WeeChat is crashing (SIGSEGV received)
*** Full crash dump was saved to /Users/Myuser/.weechat/weechat_crash_20150911_3845.log file.
***
*** Please help WeeChat developers to fix this bug:
***   1. If you have a core file, please run:  gdb /path/to/weechat core
***      then issue command: "bt full" and send result to developers
***      (see user's guide for more info about report of crashes).
***   2. Otherwise send backtrace (below), only if it is a complete trace.
***      Keep the crash log file, just in case developers ask you some info
***      (be careful, private info like passwords may be in this file).

======= WeeChat backtrace =======
(written by WeeChat 1.3, compiled on Sep  7 2015 12:26:47)
======= End of  backtrace =======
Abort trap: 6

I don't really understand which 'core' does it say... the log is pretty long (github doesn't allow to post it)
Every time I exit it creates new log file so my **~/.weechat** folder full of crash logs lol
Any ideas?
```

**Comment by flashcode** [comment]
It could be a problem with ruby plugin (which is supposed to be fixed, but on OS X I think it can still crash sometimes).
Can you please follow instructions in user's guide to send the backtrace with gdb? See here: https://weechat.org/doc/user#report_crashes
The crash file itself is useless 99% of times, because the trace is often incomplete or missing (your case).

**Comment by Drewshg312** [comment]
well first of all my weechat for some reason ignores most  of `/set` commands so after each `/quit` which ends up with crash it doesn't save the options so I'm editing weechat.conf manually (Is it normal behavior or it just can't save the changes because of crash-on-quit?). After I've done changes to activate `core` in **.weechat.conf:**

    sys_rlimit = "core:-1"

It still doesn't create **core.12345** file (neither in current dir nor in .weechat dir)
Why?
but crash report now a little different (says **core dumped**):

```
======= WeeChat backtrace =======
(written by WeeChat 1.3, compiled on Sep  7 2015 12:26:47)
======= End of  backtrace =======
Abort trap: 6 (core dumped)

So where can I find the core file?? Any ideas??

**Comment by Mikaela** [comment]
> It still doesn't create core.12345 file (neither in current dir nor in .weechat dir)
Why?
but crash report now a little different (says core dumped):

On Linux the filename is controlled by `kernel.core_pattern`, but I have no idea about OS X. Usually the coredump is created to `$HOME` and the default filename is commonly just `core`.

**Comment by Drewshg312** [comment]
well, like I said no any file with *core* pattern at all… neither in $HOME (((
Strange… may be need to set more options?

**Comment by flashcode** [comment]
The core dump should be created in the directory where you are when you run weechat (**not** the weechat install path).

**Comment by Drewshg312** [comment]
Like I said I've checked all the possibilities - I haven't found the core file neither in $(pwd) nor in $HOME or %h… It just wasn't created I believe (dunno how to check it)

**Comment by Drewshg312** [comment]
Well! Finally I've found the core file… on OSX it is in **/cores/** (NOT in the $(pwd))
But when I run `gdb` got this:

```
GNU gdb (GDB) 7.10
Copyright (C) 2015 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-apple-darwin14.5.0".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from /usr/local/bin/weechat...(no debugging symbols found)...done.
"/cores/core.10385": no core file handler recognizes format
(gdb)


Any ideas?

**Comment by flashcode** [comment]
Why is everything so different with Apple products?
I have no idea why gdb can't read a core file produced on the same OS...

And by the way, WeeChat should be compiled with debug symbols to get a trace (see user's guide for compiling it with debug symbols).

**Comment by Drewshg312** [comment]
hmm… I've installed gdb and weechat as a packages from homebrew… Will check if debug option is enabled…

**Comment by Mikaela** [comment]
> I have no idea why gdb can't read a core file produced on the same OS...

I guess it has something to do with https://github.com/Homebrew/homebrew-dupes/blob/ad263c00c2fe8d7bf418532d23492da226d4a8a6/gdb.rb#L72-L76 .

**Comment by Drewshg312** [comment]
Here we go:
http://stackoverflow.com/questions/27680712/homebrew-gdb-cant-open-core-file-on-yosemite-10-10

**Comment by Drewshg312** [comment]
Do you guys have any ideas how to do this with `lldb` instead of `gdb`? I've tried:
```
$ lldb /usr/local/bin/weechat /cores/core.70087
(lldb) target create "/usr/local/bin/weechat"
Current executable set to '/usr/local/bin/weechat' (x86_64).
(lldb) settings set -- target.run-args  "/cores/core.70087"
(lldb) bt full
error: bt [<digit>|all]
(lldb) bt all
error: invalid process
(lldb) bt
error: invalid process
```
Have no idea what am I doing (lol), just trying to do whatever I can to address these crashes… 
Please help

**Comment by Drewshg312** [comment]
I've finally found the way how to do it with lldb instead of gdb:

```
lldb --core "/cores/core.26956"                                                           ‹ruby:2.1.6› ◉
(lldb) target create --core "/cores/core.26956"
warning: (x86_64) /cores/core.26956 load command 212 LC_SEGMENT_64 has a fileoff + filesize (0x22bf8000) that extends beyond the end of the file (0x22bf7000), the segment will be truncated to match
warning: (x86_64) /cores/core.26956 load command 213 LC_SEGMENT_64 has a fileoff (0x22bf8000) that extends beyond the end of the file (0x22bf7000), ignoring this section
Core file '/cores/core.26956' (x86_64) was loaded.
(lldb) bt all
* thread #1: tid = 0x0000, 0x00007fff89976286 libsystem_kernel.dylib`__pthread_kill + 10, stop reason = signal SIGSTOP
  * frame #0: 0x00007fff89976286 libsystem_kernel.dylib`__pthread_kill + 10
    frame #1: 0x00007fff926c99f9 libsystem_pthread.dylib`pthread_kill + 90
    frame #2: 0x00007fff8b31e9b3 libsystem_c.dylib`abort + 129
    frame #3: 0x000000010bbf3a1e weechat`weechat_shutdown + 118
    frame #4: 0x00007fff8ac2af1a libsystem_platform.dylib`_sigtramp + 26
    frame #5: 0x000000010c296f58
    frame #6: 0x00007fff6ee40fbc dyld`dyld::garbageCollectImages() + 823
    frame #7: 0x00007fff6ee483e8 dyld`dlclose + 134
    frame #8: 0x00007fff91369808 libdyld.dylib`dlclose + 61
    frame #9: 0x000000010bbf041d weechat`plugin_remove + 189
    frame #10: 0x000000010bbf0784 weechat`plugin_unload + 70
    frame #11: 0x000000010bbf0868 weechat`plugin_unload_all + 40
    frame #12: 0x000000010bbf0b09 weechat`plugin_end + 16
    frame #13: 0x000000010bbf4222 weechat`weechat_end + 23
    frame #14: 0x000000010bbc8f8b weechat`main + 35
    frame #15: 0x00007fff9136a5c9 libdyld.dylib`start + 1

  thread #2: tid = 0x0001, 0x00007fff89976136 libsystem_kernel.dylib`__psynch_cvwait + 10, stop reason = signal SIGSTOP
    frame #0: 0x00007fff89976136 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fff926c8560 libsystem_pthread.dylib`_pthread_cond_wait + 693
    frame #2: 0x000000010c2a4266
    frame #3: 0x000000010c29c517
    frame #4: 0x000000010c2a2ebc
    frame #5: 0x00007fff926c805a libsystem_pthread.dylib`_pthread_body + 131
    frame #6: 0x00007fff926c7fd7 libsystem_pthread.dylib`_pthread_start + 176
    frame #7: 0x00007fff926c53ed libsystem_pthread.dylib`thread_start + 13

  thread #3: tid = 0x0002, 0x00007fff89976136 libsystem_kernel.dylib`__psynch_cvwait + 10, stop reason = signal SIGSTOP
    frame #0: 0x00007fff89976136 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fff926c8560 libsystem_pthread.dylib`_pthread_cond_wait + 693
    frame #2: 0x000000010c2a4266
    frame #3: 0x000000010c29c517
    frame #4: 0x000000010c2a2ebc
    frame #5: 0x00007fff926c805a libsystem_pthread.dylib`_pthread_body + 131
    frame #6: 0x00007fff926c7fd7 libsystem_pthread.dylib`_pthread_start + 176
    frame #7: 0x00007fff926c53ed libsystem_pthread.dylib`thread_start + 13

  thread #4: tid = 0x0003, 0x00007fff89976136 libsystem_kernel.dylib`__psynch_cvwait + 10, stop reason = signal SIGSTOP
    frame #0: 0x00007fff89976136 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fff926c8560 libsystem_pthread.dylib`_pthread_cond_wait + 693
    frame #2: 0x000000010c2a4266
    frame #3: 0x000000010c29c517
    frame #4: 0x000000010c2a2ebc
    frame #5: 0x00007fff926c805a libsystem_pthread.dylib`_pthread_body + 131
    frame #6: 0x00007fff926c7fd7 libsystem_pthread.dylib`_pthread_start + 176
    frame #7: 0x00007fff926c53ed libsystem_pthread.dylib`thread_start + 13

  thread #5: tid = 0x0004, 0x00007fff89976136 libsystem_kernel.dylib`__psynch_cvwait + 10, stop reason = signal SIGSTOP
    frame #0: 0x00007fff89976136 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fff926c8560 libsystem_pthread.dylib`_pthread_cond_wait + 693
    frame #2: 0x000000010c2a4266
    frame #3: 0x000000010c29c517
    frame #4: 0x000000010c2a2ebc
    frame #5: 0x00007fff926c805a libsystem_pthread.dylib`_pthread_body + 131
    frame #6: 0x00007fff926c7fd7 libsystem_pthread.dylib`_pthread_start + 176
    frame #7: 0x00007fff926c53ed libsystem_pthread.dylib`thread_start + 13

  thread #6: tid = 0x0005, 0x00007fff89976136 libsystem_kernel.dylib`__psynch_cvwait + 10, stop reason = signal SIGSTOP
    frame #0: 0x00007fff89976136 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fff926c8560 libsystem_pthread.dylib`_pthread_cond_wait + 693
    frame #2: 0x000000010c2a4266
    frame #3: 0x000000010c29c517
    frame #4: 0x000000010c2a2ebc
    frame #5: 0x00007fff926c805a libsystem_pthread.dylib`_pthread_body + 131
    frame #6: 0x00007fff926c7fd7 libsystem_pthread.dylib`_pthread_start + 176
    frame #7: 0x00007fff926c53ed libsystem_pthread.dylib`thread_start + 13

  thread #7: tid = 0x0006, 0x00007fff89976136 libsystem_kernel.dylib`__psynch_cvwait + 10, stop reason = signal SIGSTOP
    frame #0: 0x00007fff89976136 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fff926c8560 libsystem_pthread.dylib`_pthread_cond_wait + 693
    frame #2: 0x000000010c2a4266
    frame #3: 0x000000010c29c517
    frame #4: 0x000000010c2a2ebc
    frame #5: 0x00007fff926c805a libsystem_pthread.dylib`_pthread_body + 131
    frame #6: 0x00007fff926c7fd7 libsystem_pthread.dylib`_pthread_start + 176
    frame #7: 0x00007fff926c53ed libsystem_pthread.dylib`thread_start + 13

  thread #8: tid = 0x0007, 0x00007fff89976136 libsystem_kernel.dylib`__psynch_cvwait + 10, stop reason = signal SIGSTOP
    frame #0: 0x00007fff89976136 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fff926c8560 libsystem_pthread.dylib`_pthread_cond_wait + 693
    frame #2: 0x000000010c2a4266
    frame #3: 0x000000010c29c517
    frame #4: 0x000000010c2a2ebc
    frame #5: 0x00007fff926c805a libsystem_pthread.dylib`_pthread_body + 131
    frame #6: 0x00007fff926c7fd7 libsystem_pthread.dylib`_pthread_start + 176
    frame #7: 0x00007fff926c53ed libsystem_pthread.dylib`thread_start + 13

  thread #9: tid = 0x0008, 0x00007fff89976136 libsystem_kernel.dylib`__psynch_cvwait + 10, stop reason = signal SIGSTOP
    frame #0: 0x00007fff89976136 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fff926c8560 libsystem_pthread.dylib`_pthread_cond_wait + 693
    frame #2: 0x000000010c2a4266
    frame #3: 0x000000010c29c517
    frame #4: 0x000000010c2a2ebc
    frame #5: 0x00007fff926c805a libsystem_pthread.dylib`_pthread_body + 131
    frame #6: 0x00007fff926c7fd7 libsystem_pthread.dylib`_pthread_start + 176
    frame #7: 0x00007fff926c53ed libsystem_pthread.dylib`thread_start + 13

  thread #10: tid = 0x0009, 0x00007fff89976136 libsystem_kernel.dylib`__psynch_cvwait + 10, stop reason = signal SIGSTOP
    frame #0: 0x00007fff89976136 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fff926c8560 libsystem_pthread.dylib`_pthread_cond_wait + 693
    frame #2: 0x000000010c2a4266
    frame #3: 0x000000010c29c517
    frame #4: 0x000000010c2a2ebc
    frame #5: 0x00007fff926c805a libsystem_pthread.dylib`_pthread_body + 131
    frame #6: 0x00007fff926c7fd7 libsystem_pthread.dylib`_pthread_start + 176
    frame #7: 0x00007fff926c53ed libsystem_pthread.dylib`thread_start + 13

  thread #11: tid = 0x000a, 0x00007fff89976136 libsystem_kernel.dylib`__psynch_cvwait + 10, stop reason = signal SIGSTOP
    frame #0: 0x00007fff89976136 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fff926c8560 libsystem_pthread.dylib`_pthread_cond_wait + 693
    frame #2: 0x000000010c2a4266
    frame #3: 0x000000010c29c517
    frame #4: 0x000000010c2a2ebc
    frame #5: 0x00007fff926c805a libsystem_pthread.dylib`_pthread_body + 131
    frame #6: 0x00007fff926c7fd7 libsystem_pthread.dylib`_pthread_start + 176
    frame #7: 0x00007fff926c53ed libsystem_pthread.dylib`thread_start + 13

  thread #12: tid = 0x000b, 0x00007fff89976136 libsystem_kernel.dylib`__psynch_cvwait + 10, stop reason = signal SIGSTOP
    frame #0: 0x00007fff89976136 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fff926c8560 libsystem_pthread.dylib`_pthread_cond_wait + 693
    frame #2: 0x000000010c2a4266
    frame #3: 0x000000010c29c517
    frame #4: 0x000000010c2a2ebc
    frame #5: 0x00007fff926c805a libsystem_pthread.dylib`_pthread_body + 131
    frame #6: 0x00007fff926c7fd7 libsystem_pthread.dylib`_pthread_start + 176
    frame #7: 0x00007fff926c53ed libsystem_pthread.dylib`thread_start + 13
(lldb)
```

So what should I do now?

**Comment by Mikaela** [comment]
> I've finally found the way how to do it with lldb instead of gdb:

A little offtopic, but could you pull request that to the documentation?

**Comment by Drewshg312** [comment]
created a pull request

**Comment by flashcode** [comment]
According to your trace, it's a problem when a plugin is unloaded. The trace does not say which one, so we'll have to make some tests.
You can try to prevent some plugin to load, for example ruby (which caused crashes on exit in past): edit weechat.conf (when WeeChat is NOT running), and set the "autoload" option in `[plugin]` section to the value: `"*,!ruby"`.
Then start WeeChat and quit to check.
If it is still crashing you might try to prevent other plugins to load (try `/plugin` in WeeChat to display a list of all plugins).

**Comment by Drewshg312** [comment]
@flashcode, Thank you very much for your help!
Finally I've found: `"*,!guile"` prevents from crashing… Should I create another issue?

**Comment by flashcode** [comment]
No, this issue is OK (I'll update its title if needed).
Now we have to find where it crashes. In your trace, I don't see any WeeChat filename nor line number. Do you know if with lldb there's a way to show that?
Maybe you'll have to recompile WeeChat with debug symbols.

**Comment by Drewshg312** [comment]
Well, I've recompiled weechat (Homebrew does it with `cmake`) adding `-DCMAKE_BUILD_TYPE=Debug` argument to the formula. Is that what you mean saying "recompile WeeChat with debug symbols"?

that's what I've got now:
```
lldb --core "/cores/core.90865"                                                                                                                                                                                                                                                                                                                                                                                                                ‹ruby:2.1.6› ◉
(lldb) target create --core "/cores/core.90865"
warning: (x86_64) /cores/core.90865 load command 214 LC_SEGMENT_64 has a fileoff + filesize (0x234f8000) that extends beyond the end of the file (0x234f7000), the segment will be truncated to match
warning: (x86_64) /cores/core.90865 load command 215 LC_SEGMENT_64 has a fileoff (0x234f8000) that extends beyond the end of the file (0x234f7000), ignoring this section
Core file '/cores/core.90865' (x86_64) was loaded.
(lldb) bt all
* thread #1: tid = 0x0000, 0x00007fff89976286 libsystem_kernel.dylib`__pthread_kill + 10, stop reason = signal SIGSTOP
  * frame #0: 0x00007fff89976286 libsystem_kernel.dylib`__pthread_kill + 10
    frame #1: 0x00007fff926c99f9 libsystem_pthread.dylib`pthread_kill + 90
    frame #2: 0x00007fff8b31e9b3 libsystem_c.dylib`abort + 129
    frame #3: 0x0000000105de1a1e weechat`weechat_shutdown + 118
    frame #4: 0x00007fff8ac2af1a libsystem_platform.dylib`_sigtramp + 26
    frame #5: 0x0000000106483f58
    frame #6: 0x00007fff68156fbc dyld`dyld::garbageCollectImages() + 823
    frame #7: 0x00007fff6815e3e8 dyld`dlclose + 134
    frame #8: 0x00007fff91369808 libdyld.dylib`dlclose + 61
    frame #9: 0x0000000105dde41d weechat`plugin_remove + 189
    frame #10: 0x0000000105dde784 weechat`plugin_unload + 70
    frame #11: 0x0000000105dde868 weechat`plugin_unload_all + 40
    frame #12: 0x0000000105ddeb09 weechat`plugin_end + 16
    frame #13: 0x0000000105de2222 weechat`weechat_end + 23
    frame #14: 0x0000000105db6f8b weechat`main + 35
    frame #15: 0x00007fff9136a5c9 libdyld.dylib`start + 1
    frame #16: 0x00007fff9136a5c9 libdyld.dylib`start + 1

  thread #2: tid = 0x0001, 0x00007fff89976136 libsystem_kernel.dylib`__psynch_cvwait + 10, stop reason = signal SIGSTOP
    frame #0: 0x00007fff89976136 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fff926c8560 libsystem_pthread.dylib`_pthread_cond_wait + 693
    frame #2: 0x0000000106491266
    frame #3: 0x0000000106489517
    frame #4: 0x000000010648febc
    frame #5: 0x00007fff926c805a libsystem_pthread.dylib`_pthread_body + 131
    frame #6: 0x00007fff926c7fd7 libsystem_pthread.dylib`_pthread_start + 176
    frame #7: 0x00007fff926c53ed libsystem_pthread.dylib`thread_start + 13

  thread #3: tid = 0x0002, 0x00007fff89976136 libsystem_kernel.dylib`__psynch_cvwait + 10, stop reason = signal SIGSTOP
    frame #0: 0x00007fff89976136 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fff926c8560 libsystem_pthread.dylib`_pthread_cond_wait + 693
    frame #2: 0x0000000106491266
    frame #3: 0x0000000106489517
    frame #4: 0x000000010648febc
    frame #5: 0x00007fff926c805a libsystem_pthread.dylib`_pthread_body + 131
    frame #6: 0x00007fff926c7fd7 libsystem_pthread.dylib`_pthread_start + 176
    frame #7: 0x00007fff926c53ed libsystem_pthread.dylib`thread_start + 13

  thread #4: tid = 0x0003, 0x00007fff89976136 libsystem_kernel.dylib`__psynch_cvwait + 10, stop reason = signal SIGSTOP
    frame #0: 0x00007fff89976136 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fff926c8560 libsystem_pthread.dylib`_pthread_cond_wait + 693
    frame #2: 0x0000000106491266
    frame #3: 0x0000000106489517
    frame #4: 0x000000010648febc
    frame #5: 0x00007fff926c805a libsystem_pthread.dylib`_pthread_body + 131
    frame #6: 0x00007fff926c7fd7 libsystem_pthread.dylib`_pthread_start + 176
    frame #7: 0x00007fff926c53ed libsystem_pthread.dylib`thread_start + 13

  thread #5: tid = 0x0004, 0x00007fff89976136 libsystem_kernel.dylib`__psynch_cvwait + 10, stop reason = signal SIGSTOP
    frame #0: 0x00007fff89976136 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fff926c8560 libsystem_pthread.dylib`_pthread_cond_wait + 693
    frame #2: 0x0000000106491266
    frame #3: 0x0000000106489517
    frame #4: 0x000000010648febc
    frame #5: 0x00007fff926c805a libsystem_pthread.dylib`_pthread_body + 131
    frame #6: 0x00007fff926c7fd7 libsystem_pthread.dylib`_pthread_start + 176
    frame #7: 0x00007fff926c53ed libsystem_pthread.dylib`thread_start + 13

  thread #6: tid = 0x0005, 0x00007fff89976136 libsystem_kernel.dylib`__psynch_cvwait + 10, stop reason = signal SIGSTOP
    frame #0: 0x00007fff89976136 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fff926c8560 libsystem_pthread.dylib`_pthread_cond_wait + 693
    frame #2: 0x0000000106491266
    frame #3: 0x0000000106489517
    frame #4: 0x000000010648febc
    frame #5: 0x00007fff926c805a libsystem_pthread.dylib`_pthread_body + 131
    frame #6: 0x00007fff926c7fd7 libsystem_pthread.dylib`_pthread_start + 176
    frame #7: 0x00007fff926c53ed libsystem_pthread.dylib`thread_start + 13

  thread #7: tid = 0x0006, 0x00007fff89976136 libsystem_kernel.dylib`__psynch_cvwait + 10, stop reason = signal SIGSTOP
    frame #0: 0x00007fff89976136 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fff926c8560 libsystem_pthread.dylib`_pthread_cond_wait + 693
    frame #2: 0x0000000106491266
    frame #3: 0x0000000106489517
    frame #4: 0x000000010648febc
    frame #5: 0x00007fff926c805a libsystem_pthread.dylib`_pthread_body + 131
    frame #6: 0x00007fff926c7fd7 libsystem_pthread.dylib`_pthread_start + 176
    frame #7: 0x00007fff926c53ed libsystem_pthread.dylib`thread_start + 13

  thread #8: tid = 0x0007, 0x00007fff89976136 libsystem_kernel.dylib`__psynch_cvwait + 10, stop reason = signal SIGSTOP
    frame #0: 0x00007fff89976136 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fff926c8560 libsystem_pthread.dylib`_pthread_cond_wait + 693
    frame #2: 0x0000000106491266
    frame #3: 0x0000000106489517
    frame #4: 0x000000010648febc
    frame #5: 0x00007fff926c805a libsystem_pthread.dylib`_pthread_body + 131
    frame #6: 0x00007fff926c7fd7 libsystem_pthread.dylib`_pthread_start + 176
    frame #7: 0x00007fff926c53ed libsystem_pthread.dylib`thread_start + 13

  thread #9: tid = 0x0008, 0x00007fff89976136 libsystem_kernel.dylib`__psynch_cvwait + 10, stop reason = signal SIGSTOP
    frame #0: 0x00007fff89976136 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fff926c8560 libsystem_pthread.dylib`_pthread_cond_wait + 693
    frame #2: 0x0000000106491266
    frame #3: 0x0000000106489517
    frame #4: 0x000000010648febc
    frame #5: 0x00007fff926c805a libsystem_pthread.dylib`_pthread_body + 131
    frame #6: 0x00007fff926c7fd7 libsystem_pthread.dylib`_pthread_start + 176
    frame #7: 0x00007fff926c53ed libsystem_pthread.dylib`thread_start + 13

  thread #10: tid = 0x0009, 0x00007fff89976136 libsystem_kernel.dylib`__psynch_cvwait + 10, stop reason = signal SIGSTOP
    frame #0: 0x00007fff89976136 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fff926c8560 libsystem_pthread.dylib`_pthread_cond_wait + 693
    frame #2: 0x0000000106491266
    frame #3: 0x0000000106489517
    frame #4: 0x000000010648febc
    frame #5: 0x00007fff926c805a libsystem_pthread.dylib`_pthread_body + 131
    frame #6: 0x00007fff926c7fd7 libsystem_pthread.dylib`_pthread_start + 176
    frame #7: 0x00007fff926c53ed libsystem_pthread.dylib`thread_start + 13

  thread #11: tid = 0x000a, 0x00007fff89976136 libsystem_kernel.dylib`__psynch_cvwait + 10, stop reason = signal SIGSTOP
    frame #0: 0x00007fff89976136 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fff926c8560 libsystem_pthread.dylib`_pthread_cond_wait + 693
    frame #2: 0x0000000106491266
    frame #3: 0x0000000106489517
    frame #4: 0x000000010648febc
    frame #5: 0x00007fff926c805a libsystem_pthread.dylib`_pthread_body + 131
    frame #6: 0x00007fff926c7fd7 libsystem_pthread.dylib`_pthread_start + 176
    frame #7: 0x00007fff926c53ed libsystem_pthread.dylib`thread_start + 13

  thread #12: tid = 0x000b, 0x00007fff89976136 libsystem_kernel.dylib`__psynch_cvwait + 10, stop reason = signal SIGSTOP
    frame #0: 0x00007fff89976136 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fff926c8560 libsystem_pthread.dylib`_pthread_cond_wait + 693
    frame #2: 0x0000000106491266
    frame #3: 0x0000000106489517
    frame #4: 0x000000010648febc
    frame #5: 0x00007fff926c805a libsystem_pthread.dylib`_pthread_body + 131
    frame #6: 0x00007fff926c7fd7 libsystem_pthread.dylib`_pthread_start + 176
    frame #7: 0x00007fff926c53ed libsystem_pthread.dylib`thread_start + 13
(lldb)
```

Doesn't it look the same?

**Comment by Drewshg312** [comment]
well I've tried my best:
I've recompiled weechat (Homebrew does it with cmake) adding `-DCMAKE_BUILD_TYPE=Debug` argument to the formula and overriden the `-DCMAKE_BUILD_TYPE=Release` flag as I was told here: https://github.com/Homebrew/homebrew/pull/44370. 

Is that what you mean saying "recompile WeeChat with debug symbols"?

Also enabled in **weechat.conf**:
```
[plugin]
debug = on
```
and that's what I have so far:
```
lldb --core /cores/core.67409                                                                                                                                                                                           
(lldb) target create --core "/cores/core.67409"
warning: (x86_64) /cores/core.67409 load command 208 LC_SEGMENT_64 has a fileoff + filesize (0x21238000) that extends beyond the end of the file (0x21237000), the segment will be truncated to match
warning: (x86_64) /cores/core.67409 load command 209 LC_SEGMENT_64 has a fileoff (0x21238000) that extends beyond the end of the file (0x21237000), ignoring this section
Core file '/cores/core.67409' (x86_64) was loaded.
(lldb) bt all
* thread #1: tid = 0x0000, 0x00007fff89976286 libsystem_kernel.dylib`__pthread_kill + 10, stop reason = signal SIGSTOP
  * frame #0: 0x00007fff89976286 libsystem_kernel.dylib`__pthread_kill + 10
    frame #1: 0x00007fff926c99f9 libsystem_pthread.dylib`pthread_kill + 90
    frame #2: 0x00007fff8b31e9b3 libsystem_c.dylib`abort + 129
    frame #3: 0x00000001035f6a1e weechat`weechat_shutdown + 118
    frame #4: 0x00007fff8ac2af1a libsystem_platform.dylib`_sigtramp + 26
    frame #5: 0x0000000103cb7f58
    frame #6: 0x00007fff622f7fbc dyld`dyld::garbageCollectImages() + 823
    frame #7: 0x00007fff622ff3e8 dyld`dlclose + 134
    frame #8: 0x00007fff91369808 libdyld.dylib`dlclose + 61
    frame #9: 0x00000001035f341d weechat`plugin_remove + 189
    frame #10: 0x00000001035f3784 weechat`plugin_unload + 70
    frame #11: 0x00000001035f3868 weechat`plugin_unload_all + 40
    frame #12: 0x00000001035f3b09 weechat`plugin_end + 16
    frame #13: 0x00000001035f7222 weechat`weechat_end + 23
    frame #14: 0x00000001035cbf8b weechat`main + 35
    frame #15: 0x00007fff9136a5c9 libdyld.dylib`start + 1
    frame #16: 0x00007fff9136a5c9 libdyld.dylib`start + 1

  thread #2: tid = 0x0001, 0x00007fff89976136 libsystem_kernel.dylib`__psynch_cvwait + 10, stop reason = signal SIGSTOP
    frame #0: 0x00007fff89976136 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fff926c8560 libsystem_pthread.dylib`_pthread_cond_wait + 693
    frame #2: 0x0000000103cc5266
    frame #3: 0x0000000103cbd517
    frame #4: 0x0000000103cc3ebc
    frame #5: 0x00007fff926c805a libsystem_pthread.dylib`_pthread_body + 131
    frame #6: 0x00007fff926c7fd7 libsystem_pthread.dylib`_pthread_start + 176
    frame #7: 0x00007fff926c53ed libsystem_pthread.dylib`thread_start + 13

  thread #3: tid = 0x0002, 0x00007fff89976136 libsystem_kernel.dylib`__psynch_cvwait + 10, stop reason = signal SIGSTOP
    frame #0: 0x00007fff89976136 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fff926c8560 libsystem_pthread.dylib`_pthread_cond_wait + 693
    frame #2: 0x0000000103cc5266
    frame #3: 0x0000000103cbd517
    frame #4: 0x0000000103cc3ebc
    frame #5: 0x00007fff926c805a libsystem_pthread.dylib`_pthread_body + 131
    frame #6: 0x00007fff926c7fd7 libsystem_pthread.dylib`_pthread_start + 176
    frame #7: 0x00007fff926c53ed libsystem_pthread.dylib`thread_start + 13

  thread #4: tid = 0x0003, 0x00007fff89976136 libsystem_kernel.dylib`__psynch_cvwait + 10, stop reason = signal SIGSTOP
    frame #0: 0x00007fff89976136 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fff926c8560 libsystem_pthread.dylib`_pthread_cond_wait + 693
    frame #2: 0x0000000103cc5266
    frame #3: 0x0000000103cbd517
    frame #4: 0x0000000103cc3ebc
    frame #5: 0x00007fff926c805a libsystem_pthread.dylib`_pthread_body + 131
    frame #6: 0x00007fff926c7fd7 libsystem_pthread.dylib`_pthread_start + 176
    frame #7: 0x00007fff926c53ed libsystem_pthread.dylib`thread_start + 13

  thread #5: tid = 0x0004, 0x00007fff89976136 libsystem_kernel.dylib`__psynch_cvwait + 10, stop reason = signal SIGSTOP
    frame #0: 0x00007fff89976136 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fff926c8560 libsystem_pthread.dylib`_pthread_cond_wait + 693
    frame #2: 0x0000000103cc5266
    frame #3: 0x0000000103cbd517
    frame #4: 0x0000000103cc3ebc
    frame #5: 0x00007fff926c805a libsystem_pthread.dylib`_pthread_body + 131
    frame #6: 0x00007fff926c7fd7 libsystem_pthread.dylib`_pthread_start + 176
    frame #7: 0x00007fff926c53ed libsystem_pthread.dylib`thread_start + 13

  thread #6: tid = 0x0005, 0x00007fff89976136 libsystem_kernel.dylib`__psynch_cvwait + 10, stop reason = signal SIGSTOP
    frame #0: 0x00007fff89976136 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fff926c8560 libsystem_pthread.dylib`_pthread_cond_wait + 693
    frame #2: 0x0000000103cc5266
    frame #3: 0x0000000103cbd517
    frame #4: 0x0000000103cc3ebc
    frame #5: 0x00007fff926c805a libsystem_pthread.dylib`_pthread_body + 131
    frame #6: 0x00007fff926c7fd7 libsystem_pthread.dylib`_pthread_start + 176
    frame #7: 0x00007fff926c53ed libsystem_pthread.dylib`thread_start + 13

  thread #7: tid = 0x0006, 0x00007fff89976136 libsystem_kernel.dylib`__psynch_cvwait + 10, stop reason = signal SIGSTOP
    frame #0: 0x00007fff89976136 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fff926c8560 libsystem_pthread.dylib`_pthread_cond_wait + 693
    frame #2: 0x0000000103cc5266
    frame #3: 0x0000000103cbd517
    frame #4: 0x0000000103cc3ebc
    frame #5: 0x00007fff926c805a libsystem_pthread.dylib`_pthread_body + 131
    frame #6: 0x00007fff926c7fd7 libsystem_pthread.dylib`_pthread_start + 176
    frame #7: 0x00007fff926c53ed libsystem_pthread.dylib`thread_start + 13

  thread #8: tid = 0x0007, 0x00007fff89976136 libsystem_kernel.dylib`__psynch_cvwait + 10, stop reason = signal SIGSTOP
    frame #0: 0x00007fff89976136 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fff926c8560 libsystem_pthread.dylib`_pthread_cond_wait + 693
    frame #2: 0x0000000103cc5266
    frame #3: 0x0000000103cbd517
    frame #4: 0x0000000103cc3ebc
    frame #5: 0x00007fff926c805a libsystem_pthread.dylib`_pthread_body + 131
    frame #6: 0x00007fff926c7fd7 libsystem_pthread.dylib`_pthread_start + 176
    frame #7: 0x00007fff926c53ed libsystem_pthread.dylib`thread_start + 13

  thread #9: tid = 0x0008, 0x00007fff89976136 libsystem_kernel.dylib`__psynch_cvwait + 10, stop reason = signal SIGSTOP
    frame #0: 0x00007fff89976136 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fff926c8560 libsystem_pthread.dylib`_pthread_cond_wait + 693
    frame #2: 0x0000000103cc5266
    frame #3: 0x0000000103cbd517
    frame #4: 0x0000000103cc3ebc
    frame #5: 0x00007fff926c805a libsystem_pthread.dylib`_pthread_body + 131
    frame #6: 0x00007fff926c7fd7 libsystem_pthread.dylib`_pthread_start + 176
    frame #7: 0x00007fff926c53ed libsystem_pthread.dylib`thread_start + 13

  thread #10: tid = 0x0009, 0x00007fff89976136 libsystem_kernel.dylib`__psynch_cvwait + 10, stop reason = signal SIGSTOP
    frame #0: 0x00007fff89976136 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fff926c8560 libsystem_pthread.dylib`_pthread_cond_wait + 693
    frame #2: 0x0000000103cc5266
    frame #3: 0x0000000103cbd517
    frame #4: 0x0000000103cc3ebc
    frame #5: 0x00007fff926c805a libsystem_pthread.dylib`_pthread_body + 131
    frame #6: 0x00007fff926c7fd7 libsystem_pthread.dylib`_pthread_start + 176
    frame #7: 0x00007fff926c53ed libsystem_pthread.dylib`thread_start + 13

  thread #11: tid = 0x000a, 0x00007fff89976136 libsystem_kernel.dylib`__psynch_cvwait + 10, stop reason = signal SIGSTOP
    frame #0: 0x00007fff89976136 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fff926c8560 libsystem_pthread.dylib`_pthread_cond_wait + 693
    frame #2: 0x0000000103cc5266
    frame #3: 0x0000000103cbd517
    frame #4: 0x0000000103cc3ebc
    frame #5: 0x00007fff926c805a libsystem_pthread.dylib`_pthread_body + 131
    frame #6: 0x00007fff926c7fd7 libsystem_pthread.dylib`_pthread_start + 176
    frame #7: 0x00007fff926c53ed libsystem_pthread.dylib`thread_start + 13

  thread #12: tid = 0x000b, 0x00007fff89976136 libsystem_kernel.dylib`__psynch_cvwait + 10, stop reason = signal SIGSTOP
    frame #0: 0x00007fff89976136 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fff926c8560 libsystem_pthread.dylib`_pthread_cond_wait + 693
    frame #2: 0x0000000103cc5266
    frame #3: 0x0000000103cbd517
    frame #4: 0x0000000103cc3ebc
    frame #5: 0x00007fff926c805a libsystem_pthread.dylib`_pthread_body + 131
    frame #6: 0x00007fff926c7fd7 libsystem_pthread.dylib`_pthread_start + 176
    frame #7: 0x00007fff926c53ed libsystem_pthread.dylib`thread_start + 13
(lldb)
```

**Comment by wlhlm** [comment]
I tried inserting `dlclose()` into the weechat source and noticed that the Segfault on unload occurs only after [`weechat_plugin_init()`](https://github.com/weechat/weechat/blob/152b5ec2a7b92527695e2093d5ff45488816e83d/src/plugins/guile/weechat-guile.c#L968) has been called (which in turn uses `scm_init_guile()`).

This issue seems to be specific to OSX and has nothing to do with weechat in particular. You can trigger the Segfault using this test program:

- `guile.c`:
```c
#include <libguile.h>

extern void init_guile(void);

void init_guile(void) {
    scm_init_guile();
}
```
 - compile with `clang -g -shared -o guile.so guile.c $(pkg-config --cflags --libs guile-2.0)`

- `test.c:`
```c
#include <stdio.h>
#include <dlfcn.h>

typedef void (*guile_func)(void);

int main (void) {
    void *guile = NULL;
    guile_func init_func = NULL;

    printf("Loading Guile... ");
    guile = dlopen("guile.so", RTLD_GLOBAL | RTLD_NOW); // call used by weechat
    if (!guile) printf("Error!\n");
    else printf("Done!\n");

    printf("Trying to load \"init_guile\" ");
    init_func = dlsym(guile, "init_guile");
    if (!init_func) printf("Error!\n");
    else {
        printf("Done!\n");
        printf("Trying to run \"init_guile\"\n");
        init_func();
    }

    printf("Closing Guile... ");
    dlclose(guile);
    printf("Done!\n");
}
```
 - compile with `clang -g test.c -o test`

@Drewshg312 maybe you can try to run my test program above and see if it segfaults for you as well.

**Comment by wlhlm** [comment]
[I have reported the bug upstream](https://debbugs.gnu.org/cgi/bugreport.cgi?bug=21587)

I'm not sure it it still happens on El Capitan.

Conversation on IRC ([Log](https://gnunet.org/bot/log/guile/2015-09-28#T764407)):
> *\<wlhlm> Hi, I have a problem on Mac with shared objects that are linked with libguile. Basically when I try to unload such a shared object, a segfault gets triggered on dlclose. It would be great if someone on Mac can try to reproduce it. Here's a guide: https://gist.github.com/wlhlm/a9fd6bfd3945295fa5fb*
[...]
\<mark_weaver> wlhlm: I'm not entirely sure that we support unloading shared objects linked to libguile. typically, such shared objects will add procedures and variables to guile, which I guess means that pointers from guile global data structures will point into the shared object.
\<mark_weaver> I'm also not sure how bdw-gc, the garbage collector used by libguile, copes with this. it has some magic to add all global variables to the GC roots, and that must include globals from shared objects. if the shared object is later unmapped, I don't know if it has the necessary logic to handle that properly.
*\<wlhlm> The issue originally occured with weechat, which provides a guile scripting interface via a dynamically loaded plugin (shared object).*
[...]
\<mark_weaver> **honestly, I suspect that neither bdw-gc nor libguile have been designed to be unloaded from a process.**
\<mark_weaver> ditto for libraries linked with libguile

**Comment by jameskyle** [comment]
I'm getting the same behavior in El Capitan on HEAD when the guile plugin is installed. The behavior is consistent and happens on every exit.

    *** Very bad! WeeChat is crashing (SIGSEGV received)
    *** Full crash dump was saved to /Users/jkyle/.weechat/weechat_crash_20160203_4553.log file.
    ***
    *** Please help WeeChat developers to fix this bug:
    ***   1. If you have a core file, please run:  gdb /path/to/weechat core
    ***      then issue command: "bt full" and send result to developers
    ***      (see user's guide for more info about report of crashes).
    ***   2. Otherwise send backtrace (below), only if it is a complete trace.
    ***      Keep the crash log file, just in case developers ask you some info
    ***      (be careful, private info like passwords may be in this file).
    
    ======= WeeChat backtrace =======
    (written by WeeChat 1.5-dev (git: v1.3-167-g02bcee0), compiled on Feb  3 2016 09:47:47)
    sh: addr2line: command not found
    sh: addr2line: command not found
    sh: addr2line: command not found
    004  3   ???                                 0x0000000100989ed3 0x0 + 4304969427
    005  4   ???                                 0x00007fff6df47da8 0x0 + 140735038127528
    006  5   ???                                 0x00007fff6df4f37f 0x0 + 140735038157695
    sh: addr2line: command not found
    sh: addr2line: command not found
    sh: addr2line: command not found
    sh: addr2line: command not found
    sh: addr2line: command not found
    sh: addr2line: command not found
    sh: addr2line: command not found
    sh: addr2line: command not found
    ======= End of  backtrace =======
    [1]    4553 abort      weechat

**Comment by flashcode** [comment]
@jameskyle : can you please try to get a backtrace with gdb, as explained in the user's guide here: https://weechat.org/doc/user#report_crashes
Thank you!

**Comment by sweatsuit** [comment]
I'm having the same issue. My backtrace is below. Let me know if you need additional info, happy to help. Thanks.

(lldb) bt
* thread #1: tid = 0x0000, 0x00007fff80e32002 libsystem_kernel.dylib`__pthread_kill + 10, stop reason = signal SIGSTOP
  * frame #0: 0x00007fff80e32002 libsystem_kernel.dylib`__pthread_kill + 10
    frame #1: 0x00007fff8e2b35c5 libsystem_pthread.dylib`pthread_kill + 90
    frame #2: 0x00007fff846ea6e7 libsystem_c.dylib`abort + 129
    frame #3: 0x000000010db755db weechat`weechat_shutdown + 118
    frame #4: 0x00007fff84478eaa libsystem_platform.dylib`_sigtramp + 26
    frame #5: 0x000000010e43cfb4
    frame #6: 0x00007fff5fe76da8 dyld`dyld::garbageCollectImages() + 863
    frame #7: 0x00007fff5fe7e37f dyld`dlclose + 134
    frame #8: 0x00007fff926cd74d libdyld.dylib`dlclose + 61
    frame #9: 0x000000010db71fb2 weechat`plugin_remove + 189
    frame #10: 0x000000010db72319 weechat`plugin_unload + 70
    frame #11: 0x000000010db723fd weechat`plugin_unload_all + 40
    frame #12: 0x000000010db7269e weechat`plugin_end + 16
    frame #13: 0x000000010db75e39 weechat`weechat_end + 23
    frame #14: 0x000000010db4a593 weechat`main + 35
    frame #15: 0x00007fff926ce5ad libdyld.dylib`start + 1
    frame #16: 0x00007fff926ce5ad libdyld.dylib`start + 1

**Comment by gechr** [comment]
Adding my backtrace into the mix as I assume this is the same problem:

```
› lldb --core /cores/core.98698
(lldb) target create --core "/cores/core.98698"
warning: (x86_64) /cores/core.98698 load command 191 LC_SEGMENT_64 has a fileoff + filesize (0x28974000) that extends beyond the end of the file (0x28973000), the segment will be truncated to match
warning: (x86_64) /cores/core.98698 load command 192 LC_SEGMENT_64 has a fileoff (0x28974000) that extends beyond the end of the file (0x28973000), ignoring this section
Core file '/cores/core.98698' (x86_64) was loaded.
(lldb) bt all
* thread #1: tid = 0x0000, 0x00007fff993d8002 libsystem_kernel.dylib`__pthread_kill + 10, stop reason = signal SIGSTOP
  * frame #0: 0x00007fff993d8002 libsystem_kernel.dylib`__pthread_kill + 10
    frame #1: 0x00007fff993f05c5 libsystem_pthread.dylib`pthread_kill + 90
    frame #2: 0x00007fff9729b6e7 libsystem_c.dylib`abort + 129
    frame #3: 0x00000001012cc426 weechat`weechat_shutdown + 118
    frame #4: 0x00007fff93925eaa libsystem_platform.dylib`_sigtramp + 26
    frame #5: 0x0000000101bacfb4
    frame #6: 0x00007fff6c5a1da8 dyld`dyld::garbageCollectImages() + 863
    frame #7: 0x00007fff6c5a937f dyld`dlclose + 134
    frame #8: 0x00007fff9aba374d libdyld.dylib`dlclose + 61
    frame #9: 0x00000001012c8dfd weechat`plugin_remove + 189
    frame #10: 0x00000001012c9164 weechat`plugin_unload + 70
    frame #11: 0x00000001012c9248 weechat`plugin_unload_all + 40
    frame #12: 0x00000001012c94e9 weechat`plugin_end + 16
    frame #13: 0x00000001012ccc84 weechat`weechat_end + 23
    frame #14: 0x00000001012a16f7 weechat`main + 35
    frame #15: 0x00007fff9aba45ad libdyld.dylib`start + 1
    frame #16: 0x00007fff9aba45ad libdyld.dylib`start + 1

  thread #2: tid = 0x0001, 0x00007fff993d7eb2 libsystem_kernel.dylib`__psynch_cvwait + 10, stop reason = signal SIGSTOP
    frame #0: 0x00007fff993d7eb2 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fff993ef150 libsystem_pthread.dylib`_pthread_cond_wait + 767
    frame #2: 0x0000000101bba252
    frame #3: 0x0000000101bb256a
    frame #4: 0x0000000101bb8ea0
    frame #5: 0x00007fff993eec13 libsystem_pthread.dylib`_pthread_body + 131
    frame #6: 0x00007fff993eeb90 libsystem_pthread.dylib`_pthread_start + 168
    frame #7: 0x00007fff993ec375 libsystem_pthread.dylib`thread_start + 13

  thread #3: tid = 0x0002, 0x00007fff993d7eb2 libsystem_kernel.dylib`__psynch_cvwait + 10, stop reason = signal SIGSTOP
    frame #0: 0x00007fff993d7eb2 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fff993ef150 libsystem_pthread.dylib`_pthread_cond_wait + 767
    frame #2: 0x0000000101bba252
    frame #3: 0x0000000101bb256a
    frame #4: 0x0000000101bb8ea0
    frame #5: 0x00007fff993eec13 libsystem_pthread.dylib`_pthread_body + 131
    frame #6: 0x00007fff993eeb90 libsystem_pthread.dylib`_pthread_start + 168
    frame #7: 0x00007fff993ec375 libsystem_pthread.dylib`thread_start + 13

  thread #4: tid = 0x0003, 0x00007fff993d7eb2 libsystem_kernel.dylib`__psynch_cvwait + 10, stop reason = signal SIGSTOP
    frame #0: 0x00007fff993d7eb2 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fff993ef150 libsystem_pthread.dylib`_pthread_cond_wait + 767
    frame #2: 0x0000000101bba252
    frame #3: 0x0000000101bb256a
    frame #4: 0x0000000101bb8ea0
    frame #5: 0x00007fff993eec13 libsystem_pthread.dylib`_pthread_body + 131
    frame #6: 0x00007fff993eeb90 libsystem_pthread.dylib`_pthread_start + 168
    frame #7: 0x00007fff993ec375 libsystem_pthread.dylib`thread_start + 1
```

**Comment by flashcode** [comment]
For people reporting traces under OS X, did you try the test program above, to check if you have the OS X / guile problem?
If the problem is not related to WeeChat, it may be a good idea to forward it to OS X (or other appropriate) team.

**Comment by gechr** [comment]
Yep, it would seem to be a problem with Guile (2.0.11_2) rather than WeeChat:

```
› ./test
Loading Guile... Done!
Trying to load "init_guile" Done!
Trying to run "init_guile"
fish: './test' terminated by signal SIGSEGV (Address boundary error)
```

**Comment by flashcode** [comment]
A workaround, if you don't plan to use guile (scheme) scripts, you can prevent the plugin from loading in WeeChat:

    /set weechat.plugin.autoload "*,!guile"

**Comment by Sean1708** [comment]
For those wanting to implement the workaround above, you might have to edit the `~/.weechat/weechat.conf` file yourself.

**Comment by wlhlm** [comment]
Just as an FYI, I got a [response](https://debbugs.gnu.org/cgi/bugreport.cgi?bug=21587#11) on the upstream bug:

> Guile doesn't really provide an interface to allow it to shut down
cleanly and I don't know if it can.  (I guess it could.)  However it
would take quite some amount of work and I don't see us doing it any
time soon.  In any case simply trying to dlclose() is never likely to
work, as you aren't synchronizing with Guile itself.

> Sorry for the bad news!

Just confirming the problem. Since there is no way to unload the interpreter (à la `lua_close()` or `Py_Finalize()`), guile doesn't really work with the way weechat handles plugins in general and it seems like pure luck that it has been running fine on Linux for example.

**Comment by ronin13** [comment]
This happens on Linux as well. I can repeat crash with latest git HEAD as well as well as v1.6-dev tag. Disabling guile plugin with --disable-guile in configure prevents crash.

It is crashing in a free() in gnutls during exit.

```

#0  x86_64_fallback_frame_state (context=0x7ffc0839a850, context=0x7ffc0839a850, fs=0x7ffc0839a940) at ./md-unwind-support.h:58
#1  uw_frame_state_for (context=context@entry=0x7ffc0839a850, fs=fs@entry=0x7ffc0839a940) at /build/gcc-multilib/src/gcc/libgcc/unwind-dw2.c:1249
#2  0x00007f8e365cd6d8 in _Unwind_Backtrace (trace=0x7f8e3e2ea820 <backtrace_helper>, trace_argument=0x7ffc0839ab00) at /build/gcc-multilib/src/gcc/libgcc/unwind.inc:290
#3  0x00007f8e3e2ea98f in backtrace () from /usr/lib/libc.so.6
#4  0x000000000045344d in weechat_backtrace () at wee-backtrace.c:167
#5  0x000000000042ba6d in debug_sigsegv () at wee-debug.c:181
#6  <signal handler called>
#7  0x00007f8e380f57a0 in ?? ()
#8  0x00007f8e3ec00209 in ?? () from /usr/lib/libgnutls.so.30
#9  0x00007f8e3eb54062 in ?? () from /usr/lib/libgnutls.so.30
#10 0x00007f8e3eb65d52 in gnutls_pubkey_deinit () from /usr/lib/libgnutls.so.30
#11 0x00007f8e3eb6569c in gnutls_pcert_deinit () from /usr/lib/libgnutls.so.30
#12 0x00007f8e3ebcb3f6 in ?? () from /usr/lib/libgnutls.so.30
#13 0x00007f8e3eb5e290 in gnutls_deinit () from /usr/lib/libgnutls.so.30
#14 0x00007f8e3ab2f5c6 in irc_server_close_connection (server=0x2533d20) at irc-server.c:3218
#15 0x00007f8e3ab34cf4 in irc_server_disconnect (server=0x2533d20, switch_address=0, reconnect=0) at irc-server.c:4690
#16 0x00007f8e3ab350a3 in irc_server_disconnect_all () at irc-server.c:4763
#17 0x00007f8e3aac9e9c in weechat_plugin_end (plugin=0x24c2f70) at irc.c:290
#18 0x0000000000455a73 in plugin_unload (plugin=0x24c2f70) at plugin.c:1160
#19 0x0000000000455ba9 in plugin_unload_all () at plugin.c:1212
#20 0x0000000000455edf in plugin_end () at plugin.c:1335
#21 0x0000000000405b28 in weechat_end (gui_end_cb=0x49c1fa <gui_main_end>) at weechat.c:717
#22 0x00000000004047ff in main (argc=1, argv=0x7ffc0839b7d8) at main.c:42

```

Should be simple to reproduce. 

I configured as follows:

```

Enabled features:
   Interfaces............. : ncurses
   Plugins................ : alias aspell(aspell) charset exec fifo irc logger relay script perl(5.24.0) python(2.7) ruby(2.3.0) lua(5.3.3) guile(2.0.12) trigger xfer
   Optional features...... : gnutls flock largefile backtrace
   Compile with debug..... : yes
   Compile tests.......... : no
   Man page............... : no
   Documentation.......... : no
   Certificate authorities : /etc/ssl/certs/ca-certificates.crt

Disabled features:
   - not asked: tests man doc
   - not found: tcl javascript/v8

WeeChat home directory is ~/.weechat

WeeChat will be installed in /usr/bin

configure complete, now type 'make' to build WeeChat 1.6-dev

```

Build flags: 

```

echo "$CFLAGS \n $CXXFLAGS"
-march=x86-64 -mtune=generic -O2 -pipe -fstack-protector-strong
 -march=x86-64 -mtune=generic -O2 -pipe -fstack-protector-strong
   
```

**Comment by ronin13** [comment]
Need to verify again, but removing this ifdef fixes this crash even with guile plugin.

```
diff --git a/src/plugins/guile/weechat-guile.c b/src/plugins/guile/weechat-guile.c
index b24cf72..cba761a 100644
--- a/src/plugins/guile/weechat-guile.c
+++ b/src/plugins/guile/weechat-guile.c
@@ -969,14 +969,12 @@ weechat_plugin_init (struct t_weechat_plugin *plugin, int argc, char *argv[])
 
     guile_stdout = NULL;
 
-#ifdef HAVE_GUILE_GMP_MEMORY_FUNCTIONS
     /*
      * prevent guile to use its own gmp allocator, because it can conflict
      * with other plugins using GnuTLS like relay, which can crash WeeChat
      * on unload (or exit)
      */
     scm_install_gmp_memory_functions = 0;
-#endif /* HAVE_GUILE_GMP_MEMORY_FUNCTIONS */
 
     scm_init_guile ();

``` 

I think the ifdef may be getting set in some cases, causing the conflict with gnutls and hence the crash during exit.

If we want to keep the ifdef, may be add a ifndef GNUTLS as well, to avoid conflict.

**Comment by wlhlm** [comment]
@ronin13 your problem seems different to the one originally described, which seems particular to macOS and happens within dyld trying to free guile.so's memory. Your backtrace is different than the ones reported earlier.

The patch above does not fix the problem on macOS.

**Comment by ronin13** [comment]
@wlhlm 
Thanks, the backtraces above were incomplete, and in some cases truncated (the dyld corruption may still be due to this). Nevertheless, I will report this separately then.
